### PR TITLE
Iridius - Decentralising Web3 Frontend Software

### PR DESCRIPTION
**Pull request by Zvezdin**
### Project Abstract

> Please provide a brief description of your project here summarising key points (1-2 paragraphs).
>
> If your application is a follow-up to a previous grant, please mention which one in the first line of the abstract and include a link to previous pull requests if applicable.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying?

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [ ] I have read and understood the [FAQs](https://github.com/w3f/Grants-Program/blob/master/docs/faq.md), [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/grant_guidelines_per_category.md) and [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/announcement-guidelines.md).
- [ ] A BTC, Ethereum (USDT/USDC/DAI) or Polkadot/Kusama (aUSD) address for the payment of the milestones is provided inside the application.
- [ ] I have read and acknowledge the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [ ] The software delivered for this grant will be released under an open-source license specified in the application.
- [ ] The initial PR contains only one commit (squash and force-push if needed).
- [ ] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).
- [ ] I prefer the discussion of this application to be in a private Element/Matrix channel. My username is: ``@_______``

### How Did You Hear About our grants program?

- [ ] Social Media
- [ ] Hackathon
- [ ] Personal Recommendation
- [ ] Substrate Builders Program
- [ ] Investor/VC
- [ ] Online Search
- [ ] Other: _______

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. Something to be aware of is that we usually don’t fund the research part of projects that (potentially) want to release their own token and have a business case. Furthermore, we don’t fund the deployment part of these grant projects (E.g. Launch a public test version on a Polkadot parachain). Instead we are interested in the actual technical open source software development that is useful for others in our ecosystem and can be reused by them for their projects. Could you try to update your application accordingly? Also it would be nice to have more technical details about these deliveries? For example, what smart contract language are you going to use? We would prefer ink! here.

**Comment by Zvezdin** [comment]
Hi @Noc2, thank you for your review!

TLDR; We removed some research tasks and the deployment part, and instead focused on delivering a more complete open-source implementation of Iridius, including implementing theoretical concepts useful to other projects. We added some more technical details. The budget remains unchanged. 

Explanation and justification follows. 

I do understand that the initial grant proposal included a lot of research, not all of which was necessary for the development of the software deliverables. We have updated the proposal accordingly, only leaving the minimal amount of research for the implementation. I also want to note that by “research”, we do not mean writing a research paper or putting comparable amount of effort in it - it should be seen as investing the minimal amount of effort so that we know how can we implement what we have to, and nothing more. It would be difficult to cut more research from the deliverables, particularly because Iridius wants to bring an open source implementation of novel theoretical concepts, so these concepts need to researched in some minimal level. We believe that the implementation we provide, particularly for the MPC client in Milestone 3, will be broadly useful to many Blockchain projects, and we might be the first to implement it with these security guarantees. 

As for the development tasks which don’t seem like they are part of the final product (e.g. most of Milestone 1), we note that the final work is an iteration on top of these tasks, and as such they can be considered as part of it. 

Finally, the application focuses on describing features and functionalities, because the specifics of the MPC client implementation, such as used frameworks and languages, is too uncertain at this point. It depends on what kind of open source software can we discover and reuse to help us deliver faster, and what would be best to implement the theoretical concepts we need to. In terms of the smart contract, we are more free, and are happy to use Ink! If we find a convenient alternative of The Graph to support data reading from an Ink-based smart contract.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update. I have a few additional comments: 

- I still think that milestone 1 is ideally something that you should do before requesting a grant and I’m not sure it makes sense to sign a grant at this stage for milestones 2 and 3 without first having done the necessary research. But this is my personal opinion and I will mark the application as ready for review, so that others can take a look at it. 
- What is the advantage of your solution compared to for example using IPFS (hashes) and publishing this hash either on a trusted website or as a remark on a blockchain? Or in general using  checksums like md5? 
- Milestone 1 should also have a basic testing guide. This basically ensures that we know how to test and run your delivery.

**Comment by Zvezdin** [comment]
Hi @Noc2 , thank you for your response again. Let me answer your questions:

1. Research is the novelty (and complexity) of Iridius, and the open source development is mostly an implementation. We have already spent significant time researching prior to the grant application and even have a partial demo ready, as is described [in the whitepaper](https://drive.google.com/file/d/1YQLlk-oG6kj4lyMI9zDbPOs9rozeeKTw/view), so the questions of “is it possible/feasible/doable” are solved, and we just need to figure out exactly how it should be done (finding the right frameworks with the right performance and security guarantees, the right signatures, etc. ). A grant that funds only development (clarified & planned in advance to the detail of knowing which dependencies will be used) wouldn’t help Iridius get off the ground.

2. Our solution resolves the centralisation of off-chain software (the need to trust the developer). Let’s take the examples you suggest. If a DApp uses AAVE’s approach, and distribute its website through IPFS (or otherwise publicise a checksum), then:

- This would prevent malicious actors from hijacking the server, but cannot do anything if the developer themselves publishes a malicious update/hash.
- Hash verification cannot be done automatically, natively in a browser - the user would have to somehow download the website and manually verify the checksum. Even if the website is loaded through an IPFS gateway, the user would have to trust the gateway. When it comes to Android/iOS apps, this likewise cannot be applied (whereas what we propose - verifying a signature as originated from an MPC committee, would work).
- To protect themselves from the developer, users would have to manually review every version (checksum) of the website, which most users understandably won’t do.

Iridius decentralises this by:

- Forcing software updates to pass a community review/approval phase
- Uses a Multiparty Computation-based signature scheme, instead of hashing, which resolves some man-in-the-middle attacks and allows for offline verification
- Is natively applicable to Android/iOS apps, via that same signature scheme

3. We have updated the milestone to include a guide on how to run/manually test the deliverables.

**Review by semuelle — COMMENTED** [review]
Thanks for the application, @Zvezdin. A couple of follow-up questions:

- How does hash-based verification of websites work? Is this a hash of all public assets under a URL, or assets referenced from a specific page? I couldn't find much about how `Code Verify` does it either.
- If I publish my app signed by the MPC key, that means I couldn't publish updates without Iridius review, correct? That sounds risky.
- I don't envision any team to be willing to wait for a committee to approve every tiny change. Let's say I need to fix a spelling error or quickly warn users of an exploit or that a smart contract function has been paused; does that cause your extension to show the user a warning or even block the page?
- How long do you expect a review to take? `Code Verify` works by relying on a single source of truth—the code published by WhatsApp on Cloudflare—so updating hashes is basically instantaneous. I imagine a verification by committee to potentially take days, possible stalling attempts and sabotage from competitors notwithstanding. Looking at the size of some projects and the know-how required to understand their code (even just code changes), I wonder if this is a viable process. 
- Aren't you missing all the mobile dapp browser users by relying on a browser extension? 
- I feel as if using `Fixed reviewers` could give Iridium users a false sense of security, because those reviewers could be my henchmen, eventually making users sign malicious transactions.

**Comment by Zvezdin** [comment]
Hi @semuelle ,

Thank you for your response! You’re asking the most critical questions for Iridius - and just the ones we’ve barely touched on in this grant application, because we understood that W3F primarily funds software development. I will respond to the extent in which we currently envision it, but we are happy to be challenged and find a better model together :) 

I think most of your questions would be addressed if we clarify our current idea of how the verification process would work. We want to apply the open source model to software distribution - where many people review the code for free or little incentive, but if they find a problem, they can prove it and win a huge bounty. Once the developer submits their software, they would also include a sizeable monetary stake (e.g. share of their project tokens) as a guarantee that the software is non-malicious. Anyone is able to review the submission. If they find nothing, nothing happens, and the software gets published within some timeout window (e.g. 1-2 days). If they find that it is malicious (or has a critical security problem, not just a regular bug), the reviewer would vote “No”, provide a report explaining where the problem is, and put a monetary stake to signal their commitment. From now on, the smart contract needs to figure out who’s right - the developer, or the reviewer, anyone is invited to vote on the question, also putting a monetary stake to make sure they’re committed. Luckily, we’re talking about critical security problems, so consensus should be somewhat unanimous. After a timeout and/or some threshold is reached (e.g. 2/3), the losing side loses their stake, which gets rewarded to the winning side. So, if the developer is malicious, and a reviewer points it out, the problem should be obvious enough so many people come and vote against the developer, winning their stake. If the reviewer falsely votes “No”, then it should also be obvious that there’s no critical problem and people would be incentivised to vote against the reviewer, winning their stake. To make sure people don’t get bored of reviewing non-malicious code, the developer could once in a while publish an intentionally malicious version (in a way in which they also wouldn’t lose a lot of money). We find it exciting study & model the economics so we settle on the best solution, so let us know if you’d like to include this as part of the milestones. 

In the absolute failure scenario - a malicious developer pushes an update, which gets correctly detected by a reviewer, but then the developer’s whale friends vote against the reviewer and they lose, the reviewer’s report is still made public, together with the malicious source code, and so people are warned about the threat even if the actual software gets certified. Whereas right now, the developer could just publish a malicious update, not release the source code, and users don’t have a mechanism to learn about that until someone gets burned. As an added benefit, even if the developer is never malicious, Iridius verification removes the need to trust the middlemen - server owner, domain provider, App Store, etc. 

With that said, any solution which addresses the centralisation of Web3 software encounters the fundamental trade-off between decentralisation and control/flexibility for the developer. If the problem must be addressed, then the most cleverly designed solution (hopefully us) **will** get adopted. My personal opinion (would love to have your take on this) is that I don’t see the point of a Blockchain if 80% of your users have to trust the same developer, so if this isn’t on people’s tabs right now, I think it will be eventually. 

In order to make this response even longer than the initial grant application, here’s our take on your particular questions:
1. The way our fork of Code Verify verifies right now is by computing the hash of all JavaScript assets loaded by the webpage. This could be problematic (think, CSS key loggers or HTML attacks, putting the wrong buttons with the wrong names), so the approach we want to try is closer to what [PGP Signed Pages](https://github.com/tasn/webext-signed-pages) does. This extension hashes the entire index.html, including all CSS/JS and other assets. This is not straightforward, however, because [Chrome doesn’t implement the necessary APIs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/filterResponseData), and otherwise many extensions modify the DOM, making it difficult to hash. Hence why our current PoC is with Code Verify, but we’ll probably find a way to verify the entire DOM towards Milestone 3. 
1. For mobile apps, yes (unless your users re-install the app). For websites, you can, it’ll just fail verification. It is risky in the sense that you, as a developer, give up control and have to rely on Iridius. Less risky in the sense that users do not have to centrally trust you. 
1. You could design your software s.t. non-security critical things such as notifications or other regular events can be sourced from some server so no need for a code change. Otherwise, we can make Iridius be as flexible as people want it to be in terms of centralisation risk vs developer flexibility. For example, we can only verify the JS and DOM structure, and the devs can update the style/spelling/etc, but that opens some attack vectors (think again, misleading buttons). Another idea is to have your normal unsecured interface, and as an option for your users, a minimal web interface which is Iridius-secured. The latter would get updated every once in a while (since the smart contracts interfaces rarely change anyways, I presume you wouldn’t need frequent updates). 
1. I think I addressed that above. Also given that we want to verify only off-chain software which interacts with the Blockchain or stores secrets (wallets, DApp websites, …), so your smart contract verifying an unthinkably complex Zero-Knowledge circuit is not part of that, do you think that the code will be so heavy that many people won’t have the ability to review it? 
1. For mobile DApps, the path forward we see is for them implement the verification mechanism on their frontend (or in partnership with us). The key aspect I want to point out here is that in the long-term, Iridius establishes a backend allowing MPC signing and reliable, valuable, and difficult to replicate review process and community. Hence, different software platforms would have an interest in integrating a verifier for Iridius-certified software. 
1. The `fixed reviewers` aspect is just for the initial test version and is clearly centralised. I believe the above text explains our idea better.

**Review by takahser — CHANGES_REQUESTED** [review]
Hi @Zvezdin
Thanks for your application.

Regarding M1:

Please keep in mind that we're usually not funding Solidity/EVM-based contracts, since we're not experts in the field and also are not aware of the current state of tooling. Hence, ink! smart contracts would be preferred here.

Another thing to point out is that M1 is currently very research-heavy. I'd like to note here, that we do fund research if we think, it's beneficial to the community. However, in this case I feel like it's too closely tight to the implementation of your deliveries, which is why I'm probably not going to approve M1.

Regarding M2:

> We would like to implement the contract in ink!, but it seems like we'll have to settle for EVM. In a later milestone we will be using a decentralised data provider, such as The Graph, to read data from that contract, and for the time being they support EVM contracts, but not ink!. Let us know if there's an alternative non-centralised API provider for ink!-based contracts and we can go that way.

Did you check out the following Substrate-compatible indexing solutions yet?

- [SubQuery](https://subquery.network/)
- [Hydra](https://www.joystream.org/hydra/)
- [Polkadothub Indexer](https://github.com/figment-networks/polkadothub-indexer)
- [Substrate Graph](https://github.com/playzero/substrate-graph)

Regarding M3:

> A web interface to our smart contract.

Usually we require grantees to submit some basic wireframes if they seek funding for UI development.

To wrap up, I think it's a good idea, although I agree with @semuelle that it might be hard to convince the dApp publishers to get each little code change verified by the community. Though personally, I'd be willing to give it a shot, if you address the other issues I mentioned above. Thanks!

**Comment by semuelle** [comment]
@Zvezdin, are you planning to address @takahser's comments? Otherwise we'd close this application due to inactivity.

**Comment by Zvezdin** [comment]
Hey @semuelle , thank you for the nudge and please excuse the delayed response. Thank you @takahser too for the constructive feedback. 

Given that the consensus is that the research part of M1 should be done prior to the grant application, we decided to take the time and do so. As part of this, we're also testing out the indexing solutions @takahser mentioned (thanks!) to see if they allow for the same level of security/decentralisation as The Graph, and if that's the case, then we'll update the milestones to include ink! contracts instead. We'll also include wireframes for the web frontend. 

On the non-technical side, to address the risks you correctly identified about projects finding us too difficult to use, we are talking with projects from the Polkadot ecosystem (and outside) to nail the design of the verification process, and we're currently updating the whitepaper accordingly with our findings. 

All of this will probably take some time (~2 months), so feel free to either close the PR with the option to reopen it later, or to leave it as-is and we'll update it.

**Comment by alxs** [comment]
Hi @Zvezdin, thanks for the detailed updates. We prefer to only keep active applications open so as to not clutter our PRs, so I'll go ahead and close it. When you're ready, let us know and we'll reopen it. Note that if you force push, you will have to create a new PR.
